src/app/src/main/kotlin/org/vaccineimpact/orderlyweb/db/AuthorizationRepository.kt

```diff
@@ -0,0 +1,171 @@
+package org.vaccineimpact.orderlyweb.db
+
+import org.jooq.Record
+import org.pac4j.core.authorization.generator.AuthorizationGenerator
+import org.pac4j.core.context.WebContext
+import org.pac4j.core.profile.CommonProfile
+import org.vaccineimpact.orderlyweb.db.Tables.*
+import org.vaccineimpact.orderlyweb.errors.UnknownObjectError
+import org.vaccineimpact.orderlyweb.models.Scope
+import org.vaccineimpact.orderlyweb.models.permissions.PermissionSet
+import org.vaccineimpact.orderlyweb.models.permissions.ReifiedPermission
+import org.vaccineimpact.orderlyweb.security.authorization.orderlyWebPermissions
+
+interface AuthorizationRepository
+{
+    fun ensureUserGroupHasPermission(userGroup: String, permission: ReifiedPermission)
+}
+
+class OrderlyAuthorizationRepository : AuthorizationGenerator<CommonProfile>, AuthorizationRepository
+{
+    private val ALL_GROUP_PERMISSIONS = "all_group_permissions"
+    private val PERMISSION_NAME = "permission_name"
+    private val GROUP_PERMISSION_ID = "permission_id"
+
+    override fun generate(context: WebContext?, profile: CommonProfile): CommonProfile
+    {
+        val permissions = JooqContext().use {
+
+            val abstractPermissions =
+                    it.dsl.select(ORDERLYWEB_PERMISSION.ID.`as`(PERMISSION_NAME),
+                            ORDERLYWEB_USER_GROUP_PERMISSION.ID.`as`(GROUP_PERMISSION_ID))
+                            .fromJoinPath(ORDERLYWEB_PERMISSION,
+                                    ORDERLYWEB_USER_GROUP_PERMISSION,
+                                    ORDERLYWEB_USER_GROUP,
+                                    ORDERLYWEB_USER_GROUP_USER,
+                                    ORDERLYWEB_USER)
+                            .where(ORDERLYWEB_USER.EMAIL.eq(profile.id))
+                            .asTemporaryTable(ALL_GROUP_PERMISSIONS)
+
+            val allPermissions = getAllPermissions(it, abstractPermissions)
+            PermissionSet(allPermissions)
+        }
+
+        profile.orderlyWebPermissions = permissions
+        return profile
+    }
+
+    override fun ensureUserGroupHasPermission(userGroup: String, permission: ReifiedPermission)
+    {
+        JooqContext().use {
+            it.dsl.selectFrom(ORDERLYWEB_PERMISSION)
+                    .where(ORDERLYWEB_PERMISSION.ID.eq(permission.name))
+                    .singleOrNull() ?: throw UnknownObjectError(permission.name, "permission")
+
+            val permissionsForGroup = getAllPermissionsForGroup(it, userGroup)
+
+            if (!permissionsForGroup.any { p -> p.name == permission.name })
+            {
+                it.dsl.newRecord(ORDERLYWEB_USER_GROUP_PERMISSION).apply {
+                    this.userGroup = userGroup
+                    this.permission = permission.name
+                }.insert()
```

Comment:
Oh, so if this is working, I guess the sqlite db is doing autoincrement anway?

Reply:
Yep, see above comment.